site/app/controllers/superuser/SuperuserEmailController.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+namespace app\controllers\superuser;
+
+use app\libraries\Core;
+use app\libraries\response\WebResponse;
+use app\libraries\response\MultiResponse;
+use app\libraries\response\JsonResponse;
+use app\controllers\AbstractController;
+use app\libraries\routers\AccessControl;
+use app\libraries\User;
+use app\views\superuser\SuperuserEmailView;
+use Symfony\Component\Routing\Annotation\Route;
+use app\models\SuperuserEmail;
+
+/**
+ * @AccessControl(level="SUPERUSER")
+ */
+class SuperuserEmailController extends AbstractController {
+
+    public function __construct(Core $core) {
+        parent::__construct($core);
+    }
+
+    /**
+     * @Route("/superuser/email")
+     * @return MultiResponse
+     */
+    public function showEmailPage(): MultiResponse {
+        $num_faculty = count($this->core->getQueries()->getActiveUserIds(false, false, false, false, true));
+        $num_instructor = count($this->core->getQueries()->getActiveUserIds(true, false, false, false, false));
+        $num_full_access = count($this->core->getQueries()->getActiveUserIds(true, true, false, false, false)) - $num_instructor;
+        $num_limited_access = count($this->core->getQueries()->getActiveUserIds(true, true, true, false, false)) - $num_instructor - $num_full_access;
+        $num_student = count($this->core->getQueries()->getActiveUserIds(true, true, true, true, false)) - $num_instructor - $num_full_access - $num_limited_access;
```

Comment:
While convenient to re-use the query like this, this should be using a second DB query that just does the count within the query as that'll be quicker and use less memory.

Reply:
`public function countActiveUsersByGroup(): array {
        $this->submitty_db->query(
            "WITH A as (SELECT DISTINCT ON(user_id) user_id, user_group FROM courses_users JOIN courses
            on courses.course = courses_users.course
            and courses.semester = courses_users.semester
            WHERE courses.status = 1
            ORDER BY user_id, courses_users.user_group ASC)
            SELECT A.user_group, COUNT(A.user_group) FROM A GROUP BY A.user_group"
        );
        $result = [];
        foreach ($this->submitty_db->rows() as $row) {
            if ($row['user_group'] == 1) {
                $result['instructor'] = $row['count'];
            }
            elseif ($row['user_group'] == 2) {
                $result['full_access'] = $row['count'];
            }
            elseif ($row['user_group'] == 3) {
                $result['limited_access'] = $row['count'];
            }
            else {
                $result['student'] = $row['count'];
            }
        }
        $this->submitty_db->query(
            "SELECT COUNT(user_access_level) FROM users WHERE user_access_level <= 2"
        );
        $result['faculty'] = $this->submitty_db->row()['count'];
        return $result;
    }`
This is the query I've wrote for the counting. I separated the query for faculty however, because the results are from two different tables anyway and I don't know a good way to join them for counting. How does it look?